### PR TITLE
connect handle error, account change draft

### PR DESCRIPTION
**Pull request by irubido**
ethereum.on('accountsChanged', function (accounts) {
})
should return array of the currently available accounts

https://docs.metamask.io/guide/ethereum-provider.html#methods-new-api

### CODE REVIEW

File: example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx [hunk 1]
```diff
@@ -11,14 +12,44 @@ export const MetaMaskConnector = () => {
     const installSnap = useCallback(async () => {
        const isInitiated = await installPolkadotSnap();
        if(!isInitiated) {
-           alert("Failed to install snap");
+          dispatch({type: MetamaskActions.HAS_INSTALL_FAILED, payload: true})
```
Comment (mpetrunic) [inline example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx]: You should modify code to receive error why installation failed. You can see in docs that error code 4001 means user rejected installation so you write "Please accept snap installation prompt"

File: example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx [hunk 2]
```diff
@@ -11,14 +12,44 @@ export const MetaMaskConnector = () => {
...
        } else {
            dispatch({type: MetamaskActions.SET_INSTALLED_STATUS, payload: true});
+
+           let account: any[] = window.ethereum.on('accountsChanged', function (accounts: any) {
```
Comment (mpetrunic) [inline example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx]:
```javascript
window.ethereum.on('accountsChanged', function (accounts: string[]) {
   console.log(accounts[0]);
})
```

File: example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx [hunk 3]
```diff
@@ -11,14 +12,44 @@ export const MetaMaskConnector = () => {
...
+
        }
     }, [dispatch]);
 
+    const handleClose = (event: React.SyntheticEvent | React.MouseEvent, reason?: string) => {
+        if (reason === 'clickaway') {
+          return;
+        }
+        dispatch({type: MetamaskActions.HAS_INSTALL_FAILED, payload: false})
+      };
```
Comment (mpetrunic) [inline example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx]: where did you find this code?

File: example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx [hunk 4]
```diff
@@ -11,14 +12,44 @@ export const MetaMaskConnector = () => {
...
         <div>
+            <Snackbar
+                anchorOrigin={{
+                    vertical: 'bottom',
+                    horizontal: 'left',
+                }}
+                open={state.hasPolkadotInstallFailed}
+                autoHideDuration={6000}
+                onClose={handleClose}
+                message="Failed to install snap"
```
Comment (mpetrunic) [inline example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx]: You should display reason from context

File: example/src/context/metamask.tsx [hunk 5]
```diff
@@ -3,19 +3,22 @@ import React, {createContext, Dispatch, PropsWithChildren, Reducer, useReducer}
 
 export interface MetamaskState {
     isPolkadotSnapInstalled: boolean,
-    hasMetaMask: boolean
+    hasMetaMask: boolean,
+    hasPolkadotInstallFailed: boolean
```
Comment (mpetrunic) [inline example/src/context/metamask.tsx]: you don't need this(isPolkadotSnapInstalled will be false if it failed)
You need property that will hold installation error message

File: example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx [hunk 6]
```diff
@@ -11,14 +12,54 @@ export const MetaMaskConnector = () => {
...
        const isInitiated = await installPolkadotSnap();
        if(!isInitiated) {
-           alert("Failed to install snap");
+          dispatch({type: MetamaskActions.SET_INSTALLED_STATUS, payload: {isInstalled: false, message: "Please accept snap installation prompt"}})
        } else {
-           dispatch({type: MetamaskActions.SET_INSTALLED_STATUS, payload: true});
+          dispatch({type: MetamaskActions.SET_INSTALLED_STATUS, payload: {isInstalled: true}});
+
+          window.ethereum.on('accountsChanged', function (accounts: string[]) {
+          console.log(accounts[0]);
```
Comment (mpetrunic) [inline example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx]: Ignore this for now, aparently app key is tied to metamask, not single account

File: example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx [hunk 7]
```diff
@@ -20,15 +21,50 @@ export const MetaMaskConnector = () => {
     const installSnap = useCallback(async () => {
        const isInitiated = await installPolkadotSnap();
        if(!isInitiated) {
-           alert("Failed to install snap");
+          dispatch({type: MetamaskActions.SET_INSTALLED_STATUS, payload: {isInstalled: false, message: "Please accept snap installation prompt"}})
        } else {
+          dispatch({type: MetamaskActions.SET_INSTALLED_STATUS, payload: {isInstalled: true}});
+
+           const dotAssetAdded = await addDotAsset();
```
Comment (mpetrunic) [inline example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx:28]: I don't think you merged something properly, this ain't suppose to be awaited

File: example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx [hunk 8]
```diff
@@ -20,15 +21,50 @@ export const MetaMaskConnector = () => {
...
     }, [dispatch]);
 
+    const handleClose = (event: React.SyntheticEvent | React.MouseEvent, reason?: string) => {
+        if (reason === 'clickaway') {
+          return;
+        }
+        dispatch({type: MetamaskActions.SET_INSTALLED_STATUS, payload: false})
+      };
+    
+    const handleSnackbar = (): boolean => {
```
Comment (mpetrunic) [inline example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx:44]: shouldDisplaySnackbar

File: example/src/services/metamask.ts [hunk 9]
```diff
@@ -2,7 +2,8 @@ declare global {
     interface Window {
         ethereum: {
             isMetaMask: boolean;
-            send: (params: any)=> Promise<any>
+            send: (params: any)=> Promise<{error: any}&any[]>;
```
Comment (mpetrunic) [inline example/src/services/metamask.ts:5]: I think you should leave any response as before

File: example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx [hunk 10]
```diff
@@ -21,15 +21,9 @@ export const MetaMaskConnector = () => {
...
+        dispatch({type: MetamaskActions.SET_INSTALLED_STATUS, payload: {isInstalled: false, message: "Please accept snap installation prompt"}})
        } else {
-          dispatch({type: MetamaskActions.SET_INSTALLED_STATUS, payload: {isInstalled: true}});
-
-           const dotAssetAdded = await addDotAsset();
-           if (!dotAssetAdded) {
-               dispatch({type: MetamaskActions.SET_INSTALLED_STATUS, payload: {isInstalled: false, message: "Failed to add dot asset to metamask"}})
-           }
-           addDotAsset().catch(() => alert("Failed to add dot asset"));
+           addDotAsset().catch(() => dispatch({type: MetamaskActions.SET_INSTALLED_STATUS, payload: {isInstalled: false, message: "Failed to add dot asset to metamask"}}));
```
Comment (mpetrunic) [inline example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx:26]: I think we can let this silently fail

File: example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx [hunk 11]
```diff
@@ -20,15 +21,44 @@ export const MetaMaskConnector = () => {
...
            dispatch({type: MetamaskActions.SET_INSTALLED_STATUS, payload: true});
        }
     }, [dispatch]);
 
+    const handleClose = (event: React.SyntheticEvent | React.MouseEvent, reason?: string) => {
+        if (reason === 'clickaway') {
+          return;
+        }
+        dispatch({type: MetamaskActions.SET_INSTALLED_STATUS, payload: false})
```
Comment (mpetrunic) [inline example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx:35]: you didn't update this payload

Comment (irubido) [reply example/src/containers/MetaMaskConnector/MetaMaskConnector.tsx:35]: its just a handleClose function, it sets payload to just false, without message, hides notification when X is pressed, its controled by function in MetaMaskConnector:

const shouldDisplaySnackbar = (): boolean => {
      if (!state.isPolkadotSnapInstalled.isInstalled && state.isPolkadotSnapInstalled.message) return true;
      else return false;
    }

File: example/src/context/metamask.tsx [hunk 12]
```diff
@@ -1,13 +1,21 @@
...
 
+interface IIsPolkadotSnapInstalled {
+    isInstalled: boolean
+    message: string
+}
+
 export interface MetamaskState {
-    isPolkadotSnapInstalled: boolean,
-    hasMetaMask: boolean
+    isPolkadotSnapInstalled: IIsPolkadotSnapInstalled,
```
Comment (mpetrunic) [inline example/src/context/metamask.tsx:10]: lets think about better naming.
I would suggest to name it to `polkadotSnap`

**Pull request merged**